### PR TITLE
[bitnami/keycloak] fix configMap cert directory permission issues

### LINKED ISSUE #4821: [bitnami/keycloak] cannot create directory '/opt/bitnami/keycloak/certs': Permission denied
**Which chart**:
keycloak 1.1.4

**Describe the bug**
When deploying helm chart for Keycloak with TLS enabled, the keycloak pod fails to start with the following error
```
mkdir: cannot create directory '/opt/bitnami/keycloak/certs': Permission denied
```

**To Reproduce**

1. Enable TLS settings by configuring `answers.yaml` as follows
```
auth.tls.enabled: true
auth.tls.jksSecret: keycloak-jks
auth.tls.keystorePassword: <password>
auth.tls.truststorePassword: <password>
```
2. Deploy helm chart per instructions

**Expected behavior**
Keycloak deploys with custom TLS configuration

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
Using Rancher catalog (unsure which version of Helm is under the hood)
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.8", GitCommit:"ec6eb119b81be488b030e849b9e64fda4caaf33c", GitTreeState:"clean", BuildDate:"2020-03-12T21:00:06Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.5", GitCommit:"e0fccafd69541e3750d460ba0f9743b90336f24f", GitTreeState:"clean", BuildDate:"2020-04-16T11:35:47Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**
Red Hat Enterprise Linux 7.8
Rancher Kubernetes 17.5.1
Rancher Management 2.4.3

### PR DESCRIPTION
**Pull request by jgallucci32**
Signed-off-by: jgallucci32 <john.gallucci.iv@gmail.com>

**Description of the change**
Fixes an issue with the security context of the container image not allowing certificates to be copied to the `/opt/bitnami/keycloak/certs` directory. Also fixes a mismatch in the configMap for the environment variables to use the correct cert directory (previously was using a non-existent directory). Also simplifies the deployment to use a single TLS certificate rather than having to create one per replica.

**Benefits**
Properly configures Keycloak to use a custom TLS certificate.

**Possible drawbacks**
Simplifying TLS certificate to a single cert changes the previous method of using unique certificates for each replica of a stateful set. This is an edge case and is unknown if it is actually used in Kubernetes deployments. Given this chart did not function when initially created 2 months ago, the risk is low.

**Applicable issues**

  - fixes #4821

**Additional information**

**Checklist** 
- [ ] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [ ] Variables are documented in the README.md
- [ ] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [ ] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

### CODE REVIEW

**Comment by jgallucci32** [comment]
/hold

While the chart does fix the cert directory issue, the underlying problem of the container using the TLS settings correctly does not get fixed until version `11.0.3-debian-10-r63` of the container. The current version of the container in the chart is `11.0.3-debian-10-r59`. So do I need to bump the container version to `r63` or a more current version to fully fix TLS settings?

**Comment by FraPazGal** [comment]
Hello @jgallucci32,

Looking into this, I think the option for multiple keystores is to ensure that internal cluster secure communication is possible. I think we can readily solve the permissions error when creating the directory with an initContainer and a mounted volume. We could transfer the code done in the container's command to the initContainer where we won't have any permissions-related error, what do you think?

Regarding the container's version used in the chart, our internal tools will update them before merging the PR, so you do not need to change it.

**Comment by jgallucci32** [comment]
@FraPazGal Thanks for note about chart version. I created a separate PR to remove it from the checklist since it's handled by the internal tools https://github.com/bitnami/charts/pull/5014

If all replicas used the same certificate, shouldn't internal TLS communication still function normally? How do we validate this?

If we go down this path 3 volume mounts will be requied. The initContainer will need to mount the configMap and a second shared volume between the primary container.

**Comment by FraPazGal** [comment]
Hello @jgallucci32,

I see my colleague responded to your other PR that the checklist still has value, but thanks for taking the time anyways!

> If all replicas used the same certificate, shouldn't internal TLS communication still function normally? How do we validate this?

I don't think that would be the case as each pod would have a different hostname. I''m not sure if with the default configuration the pods communicate with each other, but they would do so in a cluster configuration. I'll prefer to use 3 volumes like you said; we would solve the current issue without dropping any functionalities and keep the door open for future additions.

**Comment by jgallucci32** [comment]
@FraPazGal I updated the PR to use an initContainer to configure the certs with a shared volume which is mounted to both containers. I also updated the logic of obtaining the ID of the replica to use a Kubernetes variable rather than rely on the Helm chart to inject it during the initial deployment.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [inline bitnami/keycloak/templates/statefulset.yaml:67]: As per bitnami's policy, we have to use images created and maintained by us in our charts. Could you modify the used image here to `bitnami/minideb`? We should also include the `securityContext` and `resources` parameters for the initContainer and have this as parameters in `values.yaml`. You could use [this other chart](https://github.com/bitnami/charts/blob/638fdadaa907a36b0304d738cb30aef366cc3255/bitnami/dokuwiki/templates/deployment.yaml#L56) as a reference.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [inline bitnami/keycloak/templates/statefulset.yaml:73]: I'm not sure `$MY_INSTANCE_NAME` is equivalent to `$fullname := include "keycloak.fullname" .`. If you take a look at the helper's logic, there are cases when these 2 envs are different and this would make `$ID` have a wrong value. I think it would be best to leave this as `$fullname`, WDYT?

**Review by FraPazGal — COMMENTED** [review]
Hi @jgallucci32!

It is looking good! There is a couple of changes needed, could you please take a look at my comments?

File: bitnami/keycloak/templates/statefulset.yaml
Comment (jgallucci32) [reply bitnami/keycloak/templates/statefulset.yaml:67]: Got it. I figured as much just didn't know which image to use. Thanks for the reference.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (jgallucci32) [reply bitnami/keycloak/templates/statefulset.yaml:73]: I'll revert back to original logic to keep it simple. I was trying to remove a dependency on Helm, but if there is a potential for conflict it's not worth the refactor.

**Comment by jgallucci32** [comment]
@FraPazGal PR updated with your suggested changes.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [inline bitnami/keycloak/templates/statefulset.yaml:67]: There is an `if` missing here:
```suggestion
      initContainers: 
        {{- if .Values.initContainers }}
        {{- include "common.tplvalues.render" (dict "value" .Values.initContainers "context" $) | nindent 8 }}
        {{- end }}
        {{- if .Values.auth.tls.enabled }}
        {{- $fullname := include "keycloak.fullname" . }}
        - name: init-certs
```

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [inline bitnami/keycloak/templates/statefulset.yaml:100]: Missing `end` form the previous if:

```suggestion
            mountPath: /opt/bitnami/keycloak/certs
        {{- end }}
      {{- end }}
```

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [inline bitnami/keycloak/templates/statefulset.yaml:79]: As we are mounting `shared-certs` to `/opt/bitnami/keycloak/certs`, we no longer need to create the folder with `mkdir`.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [inline bitnami/keycloak/templates/statefulset.yaml:75]: Our `bitnami/minideb` image gives some errors if we use `sh`, please change it to `bin/bash`:

```suggestion
          - /bin/bash
          - -ec
```

File: bitnami/keycloak/values.yaml
Comment (FraPazGal) [inline bitnami/keycloak/values.yaml:95]: Please add these new parameters to the README.

**Review by FraPazGal — COMMENTED** [review]
Hello @jgallucci32,

I tested this and there are a few issues left I commented on, minor stuff mostly. 

Also, please bump the chart's version to a new minor.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (jgallucci32) [reply bitnami/keycloak/templates/statefulset.yaml:67]: Line 62 handles the logical condition for that `if` statement. Please double check, but your suggested change should not be required.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (jgallucci32) [reply bitnami/keycloak/templates/statefulset.yaml:79]: Good catch, I will remove that.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (jgallucci32) [reply bitnami/keycloak/templates/statefulset.yaml:100]: Line `100` ends the if statement in line `65` (I double checked and also counted the if statements). One thing I noticed is that some of the bitnami charts have an extra indent under the `container` and `initContainer` sections. If you look closely enough, the entire section is 2 spaces to the right of what actually gets deployed at runtime. I thought about correcting this but left it alone in case that was done on purpose.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [reply bitnami/keycloak/templates/statefulset.yaml:67]: My bad, I didn't explain myself very well. The `if` I was referring to is `{{- if .Values.initContainers }}`. As it is now, whenever `auth.tls.enabled=true`, the code `{{- include "common.tplvalues.render" (dict "value" .Values.initContainers "context" $) | nindent 8 }}` is going to be executed, which is not something necessary. IMHO, separating this in 2 `ifs` leaves a clearer code and is the default practice in most of our charts.

Comment (jgallucci32) [reply bitnami/keycloak/templates/statefulset.yaml:67]: Ah ok I see now. You are correct my logic isn't perfect there. I agree two `if` statements needed to keep the extra include from generating.

File: bitnami/keycloak/values.yaml
Comment (jgallucci32) [reply bitnami/keycloak/values.yaml:113]: This has been completed.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (FraPazGal) [inline bitnami/keycloak/templates/statefulset.yaml:73]: This helper is missing from `_helpers.tlp`. We should include the following:

```
{{/*
Return the proper Auth TLS image name
*/}}
{{- define "keycloak.auth.tls.image" -}}
{{- include "common.images.image" (dict "imageRoot" .Values.auth.tls.image "global" .Values.global) -}}
{{- end -}}
```

File: bitnami/keycloak/Chart.yaml
Comment (FraPazGal) [inline bitnami/keycloak/Chart.yaml:29]: I'm not sure whether to have this as a patch or bump the minor version instead. It is true that this is a fix for an existing functionality, but the changes may be extensive enough for a minor, WDYT?

**Review by FraPazGal — COMMENTED** [review]
Hello @jgallucci32,

Just a couple of final changes before merging, could you please take a look?

File: bitnami/keycloak/Chart.yaml
Comment (jgallucci32) [reply bitnami/keycloak/Chart.yaml:29]: I think minor bump (as-is) since the changes focus solely on the setting `.Values.auth.tls.enabled` being true and getting it to the desired functionality. No new features have been introduced here.

File: bitnami/keycloak/templates/statefulset.yaml
Comment (jgallucci32) [reply bitnami/keycloak/templates/statefulset.yaml:73]: Thank you for pointing that out. This has been comitted.

**Review by FraPazGal — APPROVED** [review]
LGTM! Thanks for the great PR!

**Pull request merged**
